### PR TITLE
Aggregate swap history

### PR DESCRIPTION
**Pull request by iceanddust**
### What does this PR do?
- Groups multi-hop swaps into a single swap object by transaction ID.
- Add an optional parameter to the `useSwapHistory` hook to control whether you want the data returned formatted (grouped by transaction ID) or not.
- Change returned token and amount keys for `fromAmount`, `fromToken`, `toAmount` and `toToken` to represent what's the origin and target of the swap

### CODE REVIEW

File: packages/hooks/src/swap/useSwapHistory.tsx [hunk 1]
```diff
@@ -66,9 +73,66 @@ async function fetchSwapHistory(urlSubgraph: string, user?: string): Promise<Swa
...
+              blockTimestamp,
+              transactionHash,
+              pool,
+              amountUSD,
+              origin,
+              transaction,
+              amount0: fromToken ? fromToken.amount.toString() : '',
+              token0: fromToken?.token,
+              amount1: toToken ? toToken.amount.toString() : '',
+              token1: toToken?.token
```
[the remark this answers is not in the document]
Comment (iceanddust) [reply packages/hooks/src/swap/useSwapHistory.tsx]: Done, changed them to `fromAmount`, `fromToken`, `toAmount` and `toToken` in both hook responses to represent more precisely which is the origin and which is the target token

[the remark this answers is not in the document]
Comment (inclinedmind) [reply packages/hooks/src/swap/useSwapHistory.tsx]: thanks 👍 all looks good to me

**Review by inclinedmind — APPROVED** [review]

File: apps/webapp/src/modules/swap/components/SwapHistory.tsx [hunk 2]
```diff
@@ -70,11 +63,11 @@ export function SwapHistory() {
                           </Td>
                         </Tr>
                       )}
-                      {rowsToShow.map((s: SwapRecord) => (
+                      {rowsToShow.map(s => (
                         <Tr key={s.transactionHash}>
```
Comment (codewiththeflowxyz) [inline apps/webapp/src/modules/swap/components/SwapHistory.tsx:67]: This key needs to be updated as transactionHash isn's unique for ungrouped swaps. If you try adding a button that toggles formatted from true to false you'll see the rows start acting weird. This has to be done in combination with the hook's queryKey.

Sth like this works.
`<Tr key={`${s.transactionHash}-${s.id}`}>`

**Review by codewiththeflowxyz — CHANGES_REQUESTED** [review]
Overall I like it and it works well. I found a bit difficult to follow the aggregation process though so maybe we can extract some of the functionality into its own functions with a descriptive name (or simply add some comments describing what you're doing).

The only reason I requested changes is because of the `queryKey` and `<Tr key={...} > ` comments.

File: apps/webapp/src/modules/swap/components/SwapHistory.tsx (hunk 2 above)
Comment (iceanddust) [reply apps/webapp/src/modules/swap/components/SwapHistory.tsx:67]: I think using `s.id` is enough. For both response types, formatted and non-formatted, it should be unique as it consists of the transaction ID + # + a number unique for each sub-transaction

File: apps/webapp/src/modules/swap/components/SwapHistory.tsx [hunk 3]
```diff
@@ -32,17 +32,10 @@ export function SwapHistory() {
...
   // TODO how should we handle this consistently across the different history tables?
-  const formatSwapAmount = (input: string): string => {
-    const parsedNumber = parseFloat(input);
-    if (isNaN(parsedNumber)) {
-      throw new Error('Invalid input: not a valid number');
-    }
-    return Math.abs(parsedNumber)
-      .toFixed(4)
-      .replace(/\.?0*$/, '');
+  const formatSwapAmount = (input: number): number | string => {
```
[the remark this answers is not in the document]
Comment (iceanddust) [reply apps/webapp/src/modules/swap/components/SwapHistory.tsx:36]: Once that's merged we can apply that function here 👍

**Comment by inclinedmind** [comment]
looks like this is ready to be merged now?

**Review by codewiththeflowxyz — APPROVED** [review]
Good job 👍 🚀

**Pull request merged**